core/node/state_keeper/src/seal_criteria/mod.rs

```diff
@@ -186,14 +186,26 @@ pub(super) trait SealCriterion: fmt::Debug + Send + Sync + 'static {
     fn should_seal(
         &self,
         config: &StateKeeperConfig,
-        block_open_timestamp_ms: u128,
         tx_count: usize,
         l1_tx_count: usize,
         block_data: &SealData,
         tx_data: &SealData,
         protocol_version: ProtocolVersionId,
     ) -> SealResolution;
 
+    /// Returns fraction of the criterion's capacity filled in the batch.
+    /// If it can't be calculated for the criterion, then it should return `None`.
+    fn capacity_filled(
```

Comment:
Bikeshedding: IMO, it would look more intuitive to return this as the extended output from `should_seal` (whether via a tuple, or a dedicated struct), so it's easier to keep calculations in sync. The only thing making this approach awkward is that in some cases (primarily `ExcludeAndSeal` resolution, I think), it would be necessary to pass the amended block data (with the transaction excluded) to the sealer again to get the new fill stats.

Reply:
The downside is that we will mix up the critical path logic with metrics which is not good. I'd rather keep it as is.

Reply:
IMO, "mixing" logic is sort of the point in this case. As it's implemented right now, there's a risk that the returned metrics will become useless because there's nothing tying the metrics to the core business logic of the seal criteria. Probably not a huge deal given that the seal criteria would be overhauled by BoojumOS anyway 🤷